Add tutorial: How to monitor a Django application with Prometheus. (#352)
timescale/docs.timescale.com-content: tutorials/tutorial-howto-monitor-django-prometheus.md

# Tutorial: How to monitor a Django application with Prometheus

> >:TOPLIST:

Reviewer: Using a `TOPLIST` requires an H3 at the beginning of the list
```suggestion
>:TOPLIST:
> ### Sections
```
or similar

Reviewer: Having a TOPLIST has not been used with other tutorials, so if we are going to start doing this, we need to add it to all tutorials.

Author: I don't see much harm in testing this structure with a few tutorials first? I don't think any user is going to fault us too much for having a table of contents sidebar in some tutorials but not others.

> - [Introduction](#introduction)
> - [Prerequisites](#prereqs)
> - [Step 0 - Set up a basic Django application (optional)](#step0)
> - [Step 1 - Export prometheus-style monitoring metrics from your
Django application](#step1)
> - [Step 2 - Point Prometheus to your Django application metrics endpoint](#step2)
> - [Step 3 - Instrument additional aspects of your application (optional)](#step3)
> - [Next Steps](#next-steps)

> ## Introduction [](introduction)

Reviewer 2: We usually do an H1 for the title, and H3 for each section header.

Author: Cool, will update. Is there a style guide somewhere?

Prometheus is an open-source systems monitoring and alerting toolkit that can be
used to easily and cheaply monitor infrastructure and applications. In this
tutorial we show how to monitor a Django application with Prometheus. (And even
if you don’t have a Django application, we include an optional step to create
one so that everyone can follow along.)

> ## Prerequisites [](prereqs)

Reviewer 2: H3

A machine with the following installed:

- Python
- [pip][get-pip]
- A locally running [Prometheus][get-prometheus] instance

> >:Tip: Note: Since machines commonly have multiple versions of Python installed, 

Reviewer: ```suggestion
>:TIP: Since machines commonly have multiple versions of Python installed, 
```
`TIP` Must be all caps.  No need to have 'Note' also

in this tutorial we will call pip using the `python -m pip [foo]` syntax instead 
of the `pip [foo]` syntax. This is to ensure that pip installs new components for
the version of Python that we are using.

> ## Step 0 - Set up a basic Django application (optional) [](step0)

Reviewer 2: H3

*(Please skip this step if you already have a Django application.)*

> #### Install [Django][get-django]

Reviewer: Links should not be in the headline, it creates an anchor symbol before the link
<img width="483" alt="Screen Shot 2020-03-31 at 9 24 35 AM" src="https://user-images.githubusercontent.com/12288273/78031374-7878dd00-7331-11ea-9028-aba3de0ab8eb.png">

```bash
python -m pip install Django
```

#### Create a template project
```bash

> cd [to the directory where you want to create the project]

Reviewer: ```suggestion
cd {your-project-directory}
```

Reviewer: Is this command really necessary?  You already state that the second command creates a directory in the current directory

Reviewer: You could simply say 'navigate to the directory where you want to create the project' before the code block, or just leave it out.

django-admin startproject mysite
```

This will create a mysite directory in your current directory, that looks
something like this ([more here][django-first-app]):

```
mysite/
    manage.py
    mysite/
        __init__.py
        settings.py
        urls.py
        asgi.py
        wsgi.py
```

#### Verify that Django is working
Change to the outer `mysite` directory, and run:

> $ python manage.py runserver

Reviewer: Place in a code block

You should see something like this:

````
Performing system checks...

System check identified no issues (0 silenced).

You have unapplied migrations; your app may not work properly until they are applied.
Run 'python manage.py migrate' to apply them.

March 10, 2020 - 15:50:53
Django version 3.0, using settings 'mysite.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
````

If that works, then visit `http://127.0.0.1:8000/` from your Web browser.
You should see a “Congratulations!” page, with a rocket taking off.

(If that didn’t work, then please take a look at the
[Django documentation][django-first-app] for troubleshooting.)

> ## Step 1 - Export prometheus-style monitoring metrics from your Django application [](step1)

Reviewer 2: H3

We will use the [django-prometheus][get-django-prometheus] package for
exporting prometheus-style monitoring metrics from our Django application.

#### Install django-prometheus
```bash
python -m pip install django-prometheus
```

> #### Modify settings.py

Reviewer: ```suggestion
#### Modify `settings.py`
```
This step and the next step can probably be combined under a 'Modify Django settings' or 'Configure Django', but if you want to keep the filename in the headline, use inline code

In settings.py, add:

```
INSTALLED_APPS = [
   ...
   'django_prometheus',
   ...
]

MIDDLEWARE = [
    'django_prometheus.middleware.PrometheusBeforeMiddleware',
    # All your other middlewares go here, including the default
    # middlewares like SessionMiddleware, CommonMiddleware,
    # CsrfViewmiddleware, SecurityMiddleware, etc.
    'django_prometheus.middleware.PrometheusAfterMiddleware',
]
```

> #### Modify urls.py

Reviewer: ```suggestion
#### Modify `urls.py`
```

Make sure you have this in the header: 
```python
from django.conf.urls import url

Then add this under urlpatterns:
urlpatterns = [
    ...
    url('', include('django_prometheus.urls')),
]
```

#### Verify that metrics are being exported

> Restart the application and curl the /metrics endpoint:

Reviewer: ```suggestion
Restart the application and curl the `/metrics` endpoint:
```

```bash
python manage.py runserver
curl localhost:8000/metrics
```

> (Alternatively, visit [http://localhost:8000/metrics][localhost-metrics]

Reviewer: Is there a reason you made this a link?  If a user clicks without having gone through the steps, it's just a broken link.

from your web browser)

You should see something like this:

```
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 11716.0
python_gc_objects_collected_total{generation="1"} 1699.0
python_gc_objects_collected_total{generation="2"} 616.0
# HELP python_gc_objects_uncollectable_total Uncollectable object found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 7020.0
python_gc_collections_total{generation="1"} 638.0
python_gc_collections_total{generation="2"} 34.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="8",patchlevel="0",version="3.8.0"} 1.0
...
```

> ## Step 2 - Point Prometheus to your Django application metrics endpoint [](step2)

Reviewer 2: H3

*(Note: This section assumes that you have a locally running Prometheus
instance.)*

#### Update `prometheus.yml`. Under `scrape_configs:`, add:

```yaml
- job_name: django
  scrape_interval: 10s
  static_configs:
  - targets:
    - localhost:8000
```

>:TIP: Replace the job_name “django” with whatever you’d like as a prefix for
your Django application metrics in Prometheus. For example, “webapp”, etc.)

#### Restart Prometheus
```bash
./prometheus --config.file=prometheus.yml
```

> #### Verify that Prometheus is scraping metrics from yourDjango application:

Reviewer: ```suggestion
#### Verify that Prometheus is scraping metrics from your Django application:
```

> Visit the [Prometheus Expression Browser][localhost-prom-browser] from your

Reviewer: You should note that this is on the user's localhost, otherwise a reader may click on it and unknowingly be directed to a broken link (if it's not up and running)

web browser. 

For example, you can visit the below page, which graphs the total number of
http requests your Django application received in the last hour:

[http://localhost:9090/graph?g0.range_input=1h&g0.stacked=1&g0.expr=django_http_requests_total_by_method_total&g0.tab=0][localhost-prom-example]

It should look something like this:

<img class="main-content__illustration" src="https://assets.iobeam.com/images/docs/screenshots-for-tutorial-django-prometheus/prom_expression_browser.png" alt="Prometheus graph of Total HTTP Requests in the last hour"/>

If you’d like to do more testing, visit your Django application several
more times and reload the Prometheus Expression Browser to confirm that it
is working. Also feel free to explore the other Django metrics that
Prometheus is collecting.

> ## Step 3 - Instrument additional aspects of your application (optional) [](#step3)

Reviewer 2: H3

> ### Step 3 - Instrument additional aspects of your application (optional) [](#step3)

Reviewer: ```suggestion
### Step 3 - Instrument additional aspects of your application (optional) [](step3)
```

[Django-prometheus][get-django-prometheus] is quite powerful, and allows
you to easily instrument additional aspects of your application, including:

* Your databases
* Your models (e.g., monitor the creation/deletion/update rate for your models)
* Your caches

> * You can even create your own custom metrics in your code.

Reviewer: This entry doesn't fit the tenses of the first three 
> allows you to easily instrument additional aspects of your application, including: You can even create your own custom metrics in your code.

Consider taking it out of the list and placing it in a line directly below

Author: Reworded.

More information on how to do all of these is
[here][get-django-prometheus-more]. 

> ## Next steps [](#next-steps)

Reviewer 2: H3
